### PR TITLE
Enable Drag to Reorder in Playlist TableView.

### PR DESCRIPTION
**Pull request by uiryuu**
- Drag files from Finder to the playlist.
- Reorder rows in playlist.
- No more click to play item in playlist, use double click instead.
- Multiple selection in playlist table view.

I deleted PlaylistDragDestView.swift and put all the stuff in the tableview instead.
Currently PlaylistTableDelegate seems like not useful at all (because it will no longer response click action), but I left it in PlaylistViewController.swift in case of future usage. If it is better to delete it I'll push a commit about it.

### CODE REVIEW

File: iina/PlaylistViewController.swift [hunk 1]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+      var fileList: [String] = []
+      var playing: Int = 0
+      for playlistItem in playerCore.info.playlist {
+        fileList.append(playlistItem.filename)
+        if playlistItem.isPlaying {
+          playing = fileList.count - 1
+        }
+      }
+
+      for i in Array(0...playlistCount).reversed() {
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: `(0...playlistCount).reversed()` should work?

File: iina/PlaylistViewController.swift [hunk 2]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+}
+
+  func tableView(_ tableView: NSTableView, acceptDrop info: NSDraggingInfo, row: Int, dropOperation: NSTableViewDropOperation) -> Bool {
+    let pasteboard = info.draggingPasteboard()
+
+    if let rowData = pasteboard.data(forType: IINAPlaylistItemType) {
+      var dataArray = NSKeyedUnarchiver.unarchiveObject(with: rowData) as! Array<IndexSet>
+      let indexSet = dataArray[0]
+
+      let playlistCount = tableView.numberOfRows - 1
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: better get from data, i.e. `playerCore.info.playlist`

File: iina/PlaylistViewController.swift [hunk 3]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+      var order: Array<Int> = Array(0...playlistCount)
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: better use `[Int]` instead of `Array<>`

File: iina/PlaylistViewController.swift [hunk 4]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+      }
+      playlistTableView.setDropRow(row, dropOperation: .above)
+      return .copy
+    }
+    if let _ = pasteboard.propertyList(forType: IINAPlaylistItemType) {
+      playlistTableView.setDropRow(row, dropOperation: .above)
+      return .move
+    }
+    return []
+}
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: indent

File: iina/PlaylistViewController.swift [hunk 5]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+      var finalRow = row
+
+      for selectedRow in indexSet.reversed() {
+        if selectedRow < row {
+          finalRow -= 1
+        }
+        order.remove(at: selectedRow)
+      }
+
+      for selectedRow in indexSet.reversed() {
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: No need to inverse twice

```swift
let reversedIndexSet = dataArray[0].reversed()

for ... in let reversedIndexSet {
  ...
}
```

File: iina/PlaylistViewController.swift [hunk 6]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+        order.insert(selectedRow, at: finalRow)
+      }
+
+      var fileList: [String] = []
+      var playing: Int = 0
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: no need for `: Int`

File: iina/PlaylistViewController.swift [hunk 7]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+      var foundPlaying: Bool = false
+
+      for i in 0...playlistCount {
+        if order[i] == playing {
+          foundPlaying = true
+          continue
+        }
+        playerCore.addToPlaylist(fileList[order[i]])
+        if !foundPlaying {
+          playerCore.mpvController.command(.playlistMove, args: [(insertPosition + 1).toStr(), insertPosition.toStr()])
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: better add a wrapper in `PlayerCore`, like `playlistMove(from:_, to:_)`

File: iina/PlaylistViewController.swift [hunk 8]
```diff
@@ -143,6 +153,151 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource {
...
+        if i >= row {
+          finalIndexSet.insert(i + added)
+        } else {
+          finalIndexSet.insert(i)
+        }
+      }
+      tableView.deselectAll(self)
+      tableView.selectRowIndexes(finalIndexSet, byExtendingSelection: false)
+
+      NotificationCenter.default.post(Notification(name: Constants.Noti.playlistChanged))
```
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: Maybe no need for this notification, since it's always posted when `playlist/count` changed

**Review by lhc70000 — CHANGES_REQUESTED** [review]
Only some suggestions on grammar. Didn't understand the algorithm yet, let me go through it tmr...

File: iina/PlaylistViewController.swift (hunk 3 above)
Comment (alejx) [inline iina/PlaylistViewController.swift]: Just curious, what's the diffierence?

Comment (lhc70000) [inline iina/PlaylistViewController.swift]: No difference, just for better formatting. `[ ]` is a syntax sugar specially designed for this, so why not?

Comment (alejx) [inline iina/PlaylistViewController.swift]: Gotcha.

File: iina/PlaylistViewController.swift (hunk 6 above)
Comment (alejx) [inline iina/PlaylistViewController.swift]: Just curious, is it supposed to be used as few as we can?

File: iina/PlaylistViewController.swift (hunk 7 above)
Comment (alejx) [inline iina/PlaylistViewController.swift]: If I can add method in `PlayerCore`, I want to add an `insertToPlaylist(path:_, to:_)` to make code in `PlaylistViewController` more clear. What do you think?

File: iina/PlaylistViewController.swift (hunk 8 above)
Comment (alejx) [inline iina/PlaylistViewController.swift]: I tried, but after dropping, the playlist view did not refresh.

File: iina/PlaylistViewController.swift (hunk 6 above)
Comment (lhc70000) [inline iina/PlaylistViewController.swift]: Sometimes it's better to clarify the type for readability, but `Int` is a simple type (that can be inferred easily by complier) and also the default type for integer literals, so better omit for simplicity.

```swift
var x = 0  // x is Int by default
var x: UInt32 = 0  // other int types
```

Comment (alejx) [inline iina/PlaylistViewController.swift]: Sounds good!

**Comment by alejx** [comment]
@lhc70000 Ready for merge.

**Review by lhc70000 — APPROVED** [review]

**Pull request merged**